File: typescript/src/openai.ts
/**
 * OpenAI client wrapper that reports metadata to chaos trace.
 *
 * Same pattern as the Anthropic wrapper. Construct with `new OpenAI({...})`,
 * adds chaostraceApiKey + callerTag args. Every call to
 * chat.completions.create() additionally fires an async metadata report.
 */
import {
  SDK_VERSION,
  newCallId,
  newRunId,
  nowIso,
  postReportAsync,
  resolveConfig,
} from './client.js'

function getOpenAIClass(): new (config: unknown) => unknown {
  try {
    // eslint-disable-next-line @typescript-eslint/no-require-imports
    const mod = require('openai')
    const cls = mod.OpenAI ?? mod.default
    if (!cls) throw new Error('OpenAI class not found in openai package')
    return cls as new (config: unknown) => unknown
  } catch {
    throw new Error(
      "chaos trace's OpenAI wrapper requires the openai package. " +
        'Install it with: npm install openai',
    )
  }
}

export interface OpenAIOptions {
  apiKey?: string
  baseURL?: string
  chaostraceApiKey?: string
  chaostraceApiUrl?: string
  callerTag?: string
  chaostraceRunId?: string
  [key: string]: unknown
}

export class OpenAI {
  constructor(options: OpenAIOptions = {}) {
    const OpenAIClass = getOpenAIClass()
    const {
      chaostraceApiKey,
      chaostraceApiUrl,
      callerTag,
      chaostraceRunId,
      ...openaiOptions
    } = options

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const inner = new OpenAIClass(openaiOptions) as any
    const config = resolveConfig(chaostraceApiKey, chaostraceApiUrl, callerTag)
    const runId = chaostraceRunId ?? newRunId()
    const tag = config.callerTag ?? 'openai-client'

    const originalCreate = inner.chat.completions.create.bind(inner.chat.completions)
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    inner.chat.completions.create = async function (params: any): Promise<any> {
      const started = Date.now()
      const response = await originalCreate(params)
      const latencyMs = Date.now() - started

      try {
        const usage = (response && response.usage) || {}
        const model = params?.model || response?.model || 'unknown'
        postReportAsync(config, {
          spec_version: '1.0',
          agent_id: tag,
          run_id: runId,
          started_at: nowIso(),
          completed_at: nowIso(),
          status: 'completed',
          model,
          tool_calls: [
            {
              tool: 'openai.chat.completions.create',
              outcome: 'succeeded',
              decided_at: nowIso(),
            },
          ],
          outputs: [],
          metadata: {
            framework: 'chaostrace-typescript',
            sdk_version: SDK_VERSION,
            latency_ms: latencyMs,
            prompt_tokens: usage.prompt_tokens ?? null,
            completion_tokens: usage.completion_tokens ?? null,
            total_tokens: usage.total_tokens ?? null,
            call_id: newCallId(),
          },
        })
      } catch {
        // Telemetry failure must never affect the customer's code path.
      }

      return response
    }

    return inner as unknown as OpenAI
  }
}

File: python/chaostrace/_openai.py
"""OpenAI client wrapper that reports metadata to chaos trace.

Same pattern as the Anthropic wrapper: subclass the real client and
override chat.completions.create() to fire a metadata report after
each call.
"""
import time
from typing import Any, Optional

try:
    import openai as _openai
except ImportError as exc:
    raise ImportError(
        "chaos trace's OpenAI wrapper requires the openai package. "
        "Install it with: pip install openai"
    ) from exc

from ._client import (
    SDK_VERSION,
    _new_call_id,
    _new_run_id,
    _now_iso,
    _post_report_async,
    _resolve_config,
)


class OpenAI(_openai.OpenAI):
    """Drop-in replacement for openai.OpenAI."""

    def __init__(
        self,
        *args: Any,
        chaostrace_api_key: Optional[str] = None,
        chaostrace_api_url: Optional[str] = None,
        caller_tag: Optional[str] = None,
        chaostrace_run_id: Optional[str] = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(*args, **kwargs)
        self._chaostrace_config = _resolve_config(
            chaostrace_api_key, chaostrace_api_url, caller_tag
        )
        self._chaostrace_run_id = chaostrace_run_id or _new_run_id()
        self._wrap_chat_completions_create()

    def _wrap_chat_completions_create(self) -> None:
        original_create = self.chat.completions.create
        config = self._chaostrace_config
        run_id = self._chaostrace_run_id
        tag = config.caller_tag or "openai-client"

        def wrapped_create(*args: Any, **kwargs: Any) -> Any:
            started = time.monotonic()
            response = original_create(*args, **kwargs)
            elapsed_ms = int((time.monotonic() - started) * 1000)

            try:
                usage = getattr(response, "usage", None)
                model = kwargs.get("model") or getattr(response, "model", None) or "unknown"
                payload = {
                    "spec_version": "1.0",
                    "agent_id": tag,
                    "run_id": run_id,
                    "started_at": _now_iso(),
                    "completed_at": _now_iso(),
                    "status": "completed",
                    "model": model,
                    "tool_calls": [
                        {
                            "tool": "openai.chat.completions.create",
                            "outcome": "succeeded",
                            "decided_at": _now_iso(),
                        }
                    ],
                    "outputs": [],
                    "metadata": {
                        "framework": "chaostrace-python",
                        "sdk_version": SDK_VERSION,
                        "latency_ms": elapsed_ms,
                        "prompt_tokens": getattr(usage, "prompt_tokens", None) if usage else None,
                        "completion_tokens": getattr(usage, "completion_tokens", None) if usage else None,
                        "total_tokens": getattr(usage, "total_tokens", None) if usage else None,
                        "call_id": _new_call_id(),
                    },
                }
                _post_report_async(config, payload)
            except Exception:
                pass

            return response

        self.chat.completions.create = wrapped_create  # type: ignore[method-assign]
